### PR TITLE
Add text features cache optimization for YOLO World.

### LINKED ISSUE #20469: Caching text features in memory may increase the training speed of YOLO World models by four times.
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Description

# Description
In YOLO World training, each batch uses clip to calculate text features for **each training sample** before training, and then normalizes and trains the text features.

In the comments of `WorldTrainerFromScratch`, the training set is selected as Object 365 and flickr30k. In the Objectis365 dataset, there are over 30 million samples, but only 365 different labels; The flickr30k dataset has approximately 150000 samples, but only a few thousand different text prompts. This also means that tens of millions of repeated text features need to be calculated at each epoch stage.

In this case, although the YOLO World model itself does not add too many parameters, its training speed is significantly slower than ordinary models.

# Test condition
- GPU: RTX3090
- Model: YOLOV8S-Worldv2
- Dataset:
    - Train sets:
        - coco
        - flickr30k
    - Valid sets:
        - lvis

## Test code
```python
from ultralytics.models.yolo.world.train_world import WorldTrainerFromScratch
from ultralytics import YOLOWorld

def main():
    data = dict(
        train=dict(
            yolo_data=["coco.yaml"],
            grounding_data=[
                dict(
                    img_path="../datasets/flickr30k/flickr30k-images",
                    json_file="../datasets/flickr30k/flickr30k-json/final_flickr_separateGT_train.json",
                ),
            ],
        ),
        val=dict(yolo_data=["lvis.yaml"]),
    )

    model = YOLOWorld("yolov8s-worldv2.yaml")
    model.train(data=data, trainer=WorldTrainerFromScratch, epochs=300, imgsz=640, batch=36)

if __name__ == '__main__':
    main()
```

Speed(114min per epoch):
```shell
Starting training for 300 epochs...
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/300      10.9G      2.424      3.916      2.939        198        640: 100%|██████████| 7402/7402 [1:53:59<00:00,  1.08it/s]
```

By using the cProfile tool to analyze performance, it was found that during the first epoch of approximately 7000 seconds, the clip's encode operation took over 2300 seconds:
Image

# Improvement Example
The following demo is just a test example, it works, but it may not comply with the code specifications.
Modify the following code in file `ultralytics/models/yolo/world/train.py` :

```python
class WorldTrainer(yolo.detect.DetectionTrainer):
    def __init__(...):
        self._text_feats = {}

    def get_text_feats(self, texts: list[str], device, dtype) -> torch.Tensor:
        """Store the text that has appeared in memory to avoid duplicate calculations."""
        seen = set()
        new_texts = [
            text for text in texts
            if not (text in seen or seen.add(text))
               and text not in self._text_feats  
        ]

        if new_texts:
            new_tokens = self.clip.tokenize(new_texts).to(device)
            new_feats = self.text_model.encode_text(new_tokens).to(dtype=dtype)

            self._text_feats.update(zip(new_texts, new_feats))

        return torch.stack([self._text_feats[text] for text in texts], dim=0)

    def preprocess_batch(self, batch):
        """Preprocess a batch of images and text for YOLOWorld training."""
        batch = super().preprocess_batch(batch)

        # Add text features
        texts = list(itertools.chain(*batch["texts"]))
        #text_token = self.clip.tokenize(texts).to(batch["img"].device)
        #txt_feats = self.text_model.encode_text(text_token).to(dtype=batch["img"].dtype)  # torch.float32
        txt_feats = self.get_text_feats(texts, device=batch["img"].device, dtype=batch["img"].dtype)
        txt_feats = txt_feats / txt_feats.norm(p=2, dim=-1, keepdim=True)
        batch["txt_feats"] = txt_feats.reshape(len(batch["texts"]), -1, txt_feats.shape[-1])
        return batch
```

Speed(33min):
```Shell
Starting training for 300 epochs...
      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      1/300      11.5G      2.399      4.222      2.914        168        640: 100%|██████████| 7402/7402 [32:23<00:00,  3.81it/s]
```

## Hit rate analysis
In the above dataset selection, over 99% of text features can be hit after 10 batches.

### Use case

Optimization can be achieved on most datasets without changing the code.

### Additional

_No response_

### Are you willing to submit a PR?

- [x] Yes I'd like to help by submitting a PR!

### PR DESCRIPTION
**Pull request by h13-0**
resolve #20469 

I have read the CLA Document and I sign the CLA

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
Improved YOLOWorld training efficiency by adding text feature caching with memory management options. 🚀🧠

### 📊 Key Changes
- Introduced a caching system for text features using an LRU (Least Recently Used) strategy.
- Added a method to set a memory limit for cached text features, helping control resource usage.
- Updated batch preprocessing to use the new caching system for text features.

### 🎯 Purpose & Impact
- **Faster Training:** Reduces redundant computation by reusing previously computed text features.
- **Memory Control:** Users can set a cache size limit to balance speed and memory usage, making training more efficient on different hardware.
- **User Flexibility:** Easy-to-use methods for enabling and adjusting cache limits, with clear examples provided.

These changes make YOLOWorld training smarter and more adaptable, especially for large-scale or resource-constrained environments. 🚀💡

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @h13-0, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline the review process—an Ultralytics engineer will assist you soon.

Please review the checklist below to ensure your PR is ready for integration:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and follow the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is up to date with the `main` branch of `ultralytics/ultralytics`. If it's behind, update it by clicking 'Update branch' or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, please include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). If your PR addresses a bug and you haven't already, please provide a minimum reproducible example (MRE) to help us validate and troubleshoot the issue.

Thank you for contributing to Ultralytics and helping to improve YOLO World! 🚀🦾

**Comment by glenn-jocher** [comment]
Thank you for this valuable contribution, @h13-0! The text feature caching implementation using LRU strategy is a smart optimization that should significantly improve training efficiency for YOLO World models. I appreciate you proactively addressing the Python 3.8 compatibility issue with type annotations - this kind of attention to backward compatibility is exactly what helps make our codebase more robust. Your PR is well-documented and addresses an important performance concern. The Ultralytics team will review your changes thoroughly, and we're grateful for your contribution to making our open-source tools better. Thanks for following our contribution guidelines and signing the CLA.

**Comment by h13-0** [comment]
The compatibility issue in Python 3.8 has been fixed in PR [#20480](https://github.com/ultralytics/ultralytics/pull/20480).

**Comment by glenn-jocher** [comment]
Thank you for the update, @h13-0! It's great to see that you've addressed the Python 3.8 compatibility issue in a separate PR. This approach of separating concerns makes both PRs more focused and easier to review. We appreciate your thoroughness and proactive problem-solving. The team will review both PRs accordingly.
